### DISCUSSION TITLE
A hook to latex-mode doesn't trigger eglot ensure

### DISCUSSION THREAD

**Discussion by maikol-solis** [joaotavora/eglot#1340] [Q&A] [answered]
Hi!

I'm using doom-emacs, and I want to trigger eglot in latex mode automatically. I have this in my config: 

```
(add-hook! '(latex-mode-hook
               markdown-mode-hook
               prog-mode-hook) #'eglot-ensure)
```

However, it doesn't work. 

Checking a little of the eglot code, my best guess is in this line it's missing `latex-mode`. This way, eglot can find the mode correctly. 

https://github.com/emacs-straight/eglot/blob/978497f4ba22373f7ef3c0ce3008d945bda98430/eglot.el#L266

Best.

**Comment by joaotavora** [comment]
`latex-mode` derived from `tex-mode` so that line shouldn't need to be there.  Please read this:

https://joaotavora.github.io/eglot/#Starting-Eglot

Especially the second part.  Try with `M-x eglot` first, and if that doesn't work, give me more information _that I can use_.  I don't use Doom, sorry :-)

**Reply by joaotavora** [reply]
The most likely cause for your problem is that Emacs cant' find the executables to any of the servers listed for `text-mode` & friends in `eglot-server-programs`.  They should be in your `exec-path` (which is derived from `PATH` when Emacs started). If you do not want them in `PATH` _and_ you want to use `eglot-ensure` ,you have to adjust `eglot-server-programs` so that it knows the full pathname to the LSP server program.

**Reply by maikol-solis** [reply]
Running `M-x eglot` works as expected. The problem is that I can't start eglot automatically when I open a .tex file.

**Reply by joaotavora** [reply]
I don't know what your configuration looks like, but look in `*Messages*` for warnings.  Then try again after setting `debug-on-error` to `t` with `M-x toggle-debug-on-error`.

**Reply by maikol-solis** [reply, marked as answer]
Using the default setting didn't trigger any warning or error. 

However, I tried this 

```
(add-to-list 'eglot-server-programs
             '(latex-mode . ("texlab")))
```

And now eglot is triggered automatically by the `add-hook` part. 

I am just random guessing here (I don't know the internals in latex-mode). But, maybe the problem is when `tex-mode` is called, the default mode for the file is `latex-mode` and at this point there isn't any server assigned for that mode. So eglot skipped it in the `eglot-server-programs` list. 

In any case, the line I showed before solves the problem.  

Thanks.

**Reply by joaotavora** [reply]
> Using the default setting didn't trigger any warning or error.

So what you meant by "it doesn't work" remains a mystery.

> I am just random guessing here

Indeed you are, and so am I given the lack of info, but I've just installed `texlab` and this runs fine:

```
src/emacs -Q --eval "(add-hook 'latex-mode-hook 'eglot-ensure)" ~/tmp/coiso.tex
```

🤷‍♂️

**Reply by maikol-solis** [reply]
> So what you meant by "it doesn't work" remains a mystery.

In my case, the problem is that `eglot-ensure` isn't trigger automatically when I put it into the hook. 

> Indeed you are, and so am I given the lack of info, but I've just installed texlab and this runs fine:

In that case, maybe is something related to doom-emacs. I will ask them if they could put this line, `(add-to-list 'eglot-server-programs '(latex-mode . ("texlab")))` into the eglot module to catch any issue with `{latex,tex}-mode`. 

Again, thanks for your support and help to figure this out. 

Best.

**Reply by joaotavora** [reply]
> In my case, the problem is that eglot-ensure isn't trigger automatically when I put it into the hook.

The only logical consequence of the interpretation of this is that the major mode hook doesn't run.  If that were true, it would be a pretty big hole in Emacs.  So I think there is some grey area on whatever you mean by "triggered".  Functions in Elisp are "called", and you can even make Emacs stop when that happens with with `M-x toggle-debug-on-entry RET <function-name> RET`

> In that case, maybe is something related to doom-emacs. I will ask them if they could put this line, (add-to-list 'eglot-server-programs '(latex-mode . ("texlab"))) into the eglot module to catch any issue with {latex,tex}-mode.

Maybe, but that's that's not catching the issue: that is hiding it.  `latex-mode` derives from `tex-mode` and this is should be enough (and it is, at least in a configuration-less Emacs)

**Reply by maikol-solis** [reply]
Yes, I understand what you mean. I don't know what's causing this in that case. 

Just a reference, `lsp-mode` calls `latex-mode` in the client's definition, and for me, it works out-the-box,

https://github.com/emacs-lsp/lsp-mode/blob/9292d7dbe14078cec5ffd5bb39a38d6204c227d6/clients/lsp-tex.el#L59-L63

And, as you said, it could be something deeper inside the `{tex,latex}-mode`.

**Reply by joaotavora** [reply]
> Just a reference, lsp-mode calls latex-mode in the client's definition, and for me, it works out-the-box,

Out of the Emacs box, Eglot works as well (as long as you have `texlab` or `digestif` installed).  It's just out of the Doom + you configuration box that it doesn't work.

lsp-mode uses a completely different mechanism where it has a separate files for each language.  Eglot is language agnostic (in line with LSP itself ).  There is no latex-specific code in it.

> And, as you said, it could be something deeper inside the {tex,latex}-mode.

I didn't say that.  And I really really doubt it.  If you want to get to the bottom of this, bisect your configuration.  

I.e. use that Emacs invocation I gave earlier (but without the -Q) and starting with an empty configuration (where it should pass) re-add parts of your configuration until it fails.  You get the idea.

**Reply by maikol-solis** [reply]
Ok perfect! I'll do it. Thanks for your time. 

Best.

**Comment by mmarshall540** [comment]
Wild guess, but did you have AUCTeX installed?  I had the same problem and found this question while looking for a solution.  At first, the only way I could get Eglot to start was to uninstall AUCTeX.  So I ended up looking through the code to find out what's going on.

Turns out that AUCTeX replaces the function cells of both `tex-mode` and `latex-mode` during Emacs startup.  The substituted versions of those modes don't run the original hooks.  AUCTeX does this using autoloads, so bisecting your config wouldn't have caught it.

However the replaced modes do run `TeX-mode-hook` (case being important).  So if you add `eglot-ensure` to `TeX-mode-hook`, Eglot starts as expected, even with AUCTeX installed.

**Reply by maikol-solis** [reply]
Yes, I have AUCTeX installed (by doom's setup). For me, I did this to make eglot work. 
```
(add-hook! '(latex-mode-hook
               markdown-mode-hook
               prog-mode-hook) #'eglot-ensure)

(add-to-list 'eglot-server-programs
             '(latex-mode . ("texlab")))
```
